dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/DefaultMergeValidator.java

```diff
@@ -0,0 +1,210 @@
+  }
+
+  /**
+   * @param sources to check
+   * @param target to check if in sources
+   * @param mergeReport to update if any error
+   * @param clazz {@link IdentifiableObject} type
+   */
+  @Override
+  public <T extends IdentifiableObject> void checkIsTargetInSources(
+      Set<UID> sources, UID target, MergeReport mergeReport, Class<T> clazz) {
+    if (sources.contains(target))
+      mergeReport.addErrorMessage(
+          new ErrorMessage(targetNotSourceErrorCode(clazz.getSimpleName())));
+  }
+
+  /**
+   * Verifies whether the target {@link UID} maps to a valid {@link IdentifiableObject}. <br>
+   * - If it's valid then a fully populated {@link MergeRequest} is returned. <br>
+   * - If it's not valid then an empty {@link MergeRequest} is returned and the {@link MergeReport}
+   * is updated with an error.
+   *
+   * @param mergeReport to update if any error
+   * @param sources to return in merge request
+   * @param params merge params with target to verify
+   * @param clazz {@link IdentifiableObject} type
+   * @return merge request
+   */
+  @Override
+  public <T extends IdentifiableObject> MergeRequest verifyTarget(
+      MergeReport mergeReport, Set<UID> sources, MergeParams params, Class<T> clazz) {
+    return getAndVerify(params.getTarget(), mergeReport, MergeObjectType.TARGET.name(), clazz)
+        .map(
+            t ->
+                MergeRequest.builder()
+                    .sources(sources)
+                    .target(t)
+                    .deleteSources(params.isDeleteSources())
+                    .build())
+        .orElse(MergeRequest.empty());
+  }
+
+  /**
+   * Verifies whether the {@link UID} maps to a valid {@link IdentifiableObject}. <br>
+   * - If a valid {@link IdentifiableObject} is found then an {@link Optional} of the {@link UID} is
+   * returned. <br>
+   * - If a valid {@link IdentifiableObject} is not found then an empty {@link Optional} is returned
+   * and an error is added to the {@link MergeReport}.
+   *
+   * @param uid to verify
+   * @param mergeReport to update
+   * @param mergeObjectType indicating whether this is a source or target
+   * @param clazz {@link IdentifiableObject} type
+   * @return optional UID
+   */
+  private <T extends IdentifiableObject> Optional<UID> getAndVerify(
+      UID uid, MergeReport mergeReport, String mergeObjectType, Class<T> clazz) {
+
+    if (uid != null) {
+      return Optional.ofNullable(manager.get(clazz, uid.getValue()))
+          .map(i -> UID.of(i.getUid()))
+          .or(
+              () -> {
+                mergeReport.addErrorMessage(
+                    new ErrorMessage(
+                        doesNotExistErrorCode(clazz.getSimpleName()), mergeObjectType, uid));
+                return Optional.empty();
+              });
+    } else {
+      mergeReport.addErrorMessage(
+          new ErrorMessage(noTargetErrorCode(clazz.getSimpleName()), mergeObjectType, uid));
+      return Optional.empty();
```

Comment:
That is a bit over my use of `Optional` limit :D

Reply:
that's a very low limit :) 
all it does is get an optional and do something if it's present or return empty if not present.
i'll extract a method so it reads a bit nicer.